examples/nlp/multi_label_classification.py

```diff
@@ -0,0 +1,382 @@
+
+    # The expansion is needed for subsequent vectorization.
+    return tf.expand_dims(unified_text, -1), label
+
+
+def make_dataset(dataframe, is_train=True):
+    label_binarized = mlb.transform(dataframe["terms"].values)
+    dataset = tf.data.Dataset.from_tensor_slices(
+        (dataframe["summaries"].values, label_binarized)
+    )
+    dataset = dataset.shuffle(batch_size * 10) if is_train else dataset
+    dataset = dataset.map(unify_text_length).cache()
+    return dataset.batch(batch_size).prefetch(tf.data.AUTOTUNE)
+
+
+"""
+Now we can prepare the `tf.data.Dataset` objects. 
+"""
+
+train_dataset = make_dataset(train_df, is_train=True)
+validation_dataset = make_dataset(val_df, is_train=False)
+test_dataset = make_dataset(test_df, is_train=False)
+
+"""
+## Dataset preview
+"""
+
+text_batch, label_batch = next(iter(train_dataset))
+
+for i, text in enumerate(text_batch[:5]):
+    label = label_batch[i].numpy()[None, ...]
+    print(f"Abstract: {text[0]}")
+    print(f"Label(s): {mlb.inverse_transform(label)[0]}")
+    print(" ")
+
+"""
+## Vocabulary size for vectorization
+
+Before we feed the data to our model we need to represent them as numbers. For that
+purpose, we will use the
+[`TextVectorization` layer](https://keras.io/api/layers/preprocessing_layers/text/text_vectorization).
+It can operate as a part of your main model so that the model is excluded from the core
+preprocessing logic. This greatly reduces the chances of training and serving skew.
+
+We first calculate the number of unique words present in the abstracts.
+"""
+
+train_df["total_words"] = train_df["summaries"].str.split().str.len()
+vocabulary_size = train_df["total_words"].max()
+print(f"Vocabulary size: {vocabulary_size}")
+
+"""
+Now we can create our text classifier model with the `TextVectorization` layer present
+inside it. 
+"""
+
+"""
+## Create model with `TextVectorization`
+
+A batch of raw text will first go through the `TextVectorization` layer and it will
+generate their integer representations. Internally, the `TextVectorization` layer will
+first create bi-grams out of the sequences and then represent them using
+[TF-IDF](https://wikipedia.org/wiki/Tf%E2%80%93idf). The output representations will then
+be passed to the shallow model responsible for text classification. 
+
+To know more about other possible configurations with `TextVectorizer`, please consult
+the 
+[official documentation](https://keras.io/api/layers/preprocessing_layers/text/text_vectorization).
+
+"""
+
+text_vectorizer = layers.TextVectorization(
+    max_tokens=vocabulary_size, ngrams=2, output_mode="tf_idf"
+)
+
+# `TextVectorization` needs to be adapted as per the vocabulary from our
+# training set.
+with tf.device("/CPU:0"):
+    text_vectorizer.adapt(train_dataset.map(lambda text, label: text))
+
+
+def make_model():
+    shallow_mlp_model = keras.Sequential(
+        [
+            text_vectorizer,
```

Comment:
Prefer `map()`ing the layer in the tf.data pipeline

Reply:
Oh, we wanted to let the model directly take raw strings hence this choice. We understand that it's totally possible to export a separate model including the `TextVectorization` layer on top of the trained model but when we wanted to show the users a different way. Is that okay?

Reply:
It's preferable to always train with a model where `TextVectorization` is in the data pipeline, to enable async preprocessing. Otherwise you can create a performance bottleneck on GPU/TPU

Reply:
Alright. 

When `map()`ing the `TextVectorization` layer what would be the preferred way? Should we do something like the following?

```
text_vectorizer.adapt(train_dataset.map(lambda text, label: text))
train_dataset = train_dataset.map(lambda text, label: (text_vectorizer(text), label))
validation_dataset = ...
test_dataset = ...
```

Note that `train_dataset` is already batched.

Reply:
Yes, this works, but add `num_parallel_calls`. Preprocessing layers work on both batched and unbatched data.